# ABC Corporation

## Imports y carga de archivos

In [1]:
# importar librerías
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
import re

In [2]:
# cargar datos
df_abc = pd.read_csv('../files/hr.csv')

## Columnas

- age: Edad del empleado.
- attrition: Indica si el empleado dejó la empresa (Yes/No).
- business_travel: Frecuencia de viajes laborales (por ejemplo, Travel_Rarely, Travel_Frequently).
- daily_rate: Tarifa diaria estimada basada en el salario.
- department: Departamento en el que trabaja el empleado.
- distance_from_home: Distancia desde el hogar hasta el lugar de trabajo.
- education: Nivel educativo del empleado representado en una escala numérica.
- education_field: Campo o área de estudios del empleado.
- employee_count: Valor constante (normalmente 1), que indica un empleado por registro.
- employee_number: Identificador único del empleado.
- environment_satisfaction: Nivel de satisfacción con el entorno laboral.
- gender: Género del empleado.
- hourly_rate: Tarifa por hora estimada.
- job_involvement: Nivel de implicación del empleado en su trabajo.
- job_level: Nivel jerárquico del puesto.
- job_role: Rol o puesto específico del empleado.
- job_satisfaction: Nivel de satisfacción con el trabajo.
- marital_status: Estado civil del empleado.
- monthly_income: Ingreso mensual estimado.
- monthly_rate: Tarifa mensual estimada.
- num_companies_worked: Número de empresas en las que ha trabajado previamente.
- over18: Indica si el empleado es mayor de 18 años.
- over_time: Indica si el empleado realiza horas extra (Yes/No).
- percent_salary_hike: Porcentaje de incremento salarial.
- performance_rating: Evaluación del desempeño del empleado.
- relationship_satisfaction: Nivel de satisfacción con las relaciones laborales.
- standard_hours: Horas estándar de trabajo.
- stock_option_level: Nivel de opciones sobre acciones asignadas.
- total_working_years: Total de años de experiencia laboral.
- training_times_last_year: Número de formaciones recibidas en el último año.
- work_life_balance: Nivel de equilibrio entre vida personal y laboral.
- years_at_company: Años que el empleado lleva en la empresa.
- years_in_current_role: Años en el rol actual.
- years_since_last_promotion: Años transcurridos desde la última promoción.
- years_with_curr_manager: Años trabajando con el gerente actual

## Exploración inicial

In [3]:
# mostrar las primeras filas del DataFrame
df_abc.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0  41.0       Yes      Travel_Rarely       1102                   Sales   
1  49.0        No  Travel_Frequently        279  Research & Development   
2  37.0       Yes      Travel_Rarely       1373  Research & Development   
3  33.0        No  Travel_Frequently       1392  Research & Development   
4  27.0        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1          80.0                 0   
1  ...                         4           NaN                 1   
2  ...                         2           NaN                 0   
3  ...                         3          80.0                 0   
4  ...                         4          80.0                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                    0.0               1               6   
1                 10                    3.0               3              10   
2                  7                    3.0               3               0   
3                  8                    3.0               3               8   
4                  6                    3.0               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                   5.0  
1                  7                        1                   7.0  
2                  0                        0                   0.0  
3                  7                        3                   0.0  
4                  2                        2                   2.0  

[5 rows x 35 columns]

In [4]:
df_abc.sample(5)

Age Attrition BusinessTravel  DailyRate              Department  \
425  50.0        No  Travel_Rarely       1099  Research & Development   
559  38.0        No  Travel_Rarely        268  Research & Development   
539  50.0        No  Travel_Rarely        316                   Sales   
596  35.0        No  Travel_Rarely       1258  Research & Development   
976  56.0        No  Travel_Rarely       1369  Research & Development   

     DistanceFromHome  Education EducationField  EmployeeCount  \
425                29          4  Life Sciences              1   
559                 2          5        Medical              1   
539                 8          4      Marketing              1   
596                 1          4  Life Sciences              1   
976                23          3  Life Sciences              1   

     EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
425             569  ...                         2          80.0   
559             773  ...                         2          80.0   
539             738  ...                         4          80.0   
596             826  ...                         3           NaN   
976            1373  ...                         1          80.0   

     StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
425                 1                 28                    2.0   
559                 1                  6                    0.0   
539                 1                  4                    2.0   
596                 0                  7                    0.0   
976                 1                 33                    0.0   

    WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
425               3              27                 10   
559               1               1                  0   
539               3               2                  2   
596               3               2                  2   
976               3              19                 16   

     YearsSinceLastPromotion  YearsWithCurrManager  
425                       15                   7.0  
559                        0                   1.0  
539                        2                   2.0  
596                        2                   2.0  
976                       15                   9.0  

[5 rows x 35 columns]

In [5]:
# mostrar las últimas filas del DataFrame
df_abc.tail()

Age Attrition BusinessTravel  DailyRate              Department  \
1469  34.0        No  Travel_Rarely        628  Research & Development   
1470  28.0        No  Travel_Rarely        866                   Sales   
1471  53.0        No  Travel_Rarely       1084  Research & Development   
1472  24.0       Yes  Travel_Rarely        240         Human Resources   
1473  45.0        No  Travel_Rarely       1339  Research & Development   

      DistanceFromHome  Education   EducationField  EmployeeCount  \
1469                 8          3          Medical              1   
1470                 5          3          Medical              1   
1471                13          2          Medical              1   
1472                22          1  Human Resources              1   
1473                 7          3    Life Sciences              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
1469            2068  ...                         1           NaN   
1470            1469  ...                         4           NaN   
1471             250  ...                         3           NaN   
1472            1714  ...                         3          80.0   
1473              86  ...                         3           NaN   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1469                 0                  6                    3.0   
1470                 0                  6                    4.0   
1471                 2                  5                    3.0   
1472                 1                  1                    2.0   
1473                 1                 25                    2.0   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
1469               4               4                  3   
1470               3               5                  4   
1471               3               4                  2   
1472               3               1                  0   
1473               3               1                  0   

      YearsSinceLastPromotion  YearsWithCurrManager  
1469                        1                   2.0  
1470                        1                   NaN  
1471                        1                   3.0  
1472                        0                   0.0  
1473                        0                   0.0  

[5 rows x 35 columns]

In [6]:
# mostrar el tamaño del DataFrame
df_abc.shape

(1474, 35)

In [7]:
# mostrar información del DataFrame
df_abc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1474 entries, 0 to 1473
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1401 non-null   float64
 1   Attrition                 1474 non-null   object 
 2   BusinessTravel            1357 non-null   object 
 3   DailyRate                 1474 non-null   int64  
 4   Department                1445 non-null   object 
 5   DistanceFromHome          1474 non-null   int64  
 6   Education                 1474 non-null   int64  
 7   EducationField            1416 non-null   object 
 8   EmployeeCount             1474 non-null   int64  
 9   EmployeeNumber            1474 non-null   int64  
 10  EnvironmentSatisfaction   1474 non-null   int64  
 11  Gender                    1474 non-null   object 
 12  HourlyRate                1474 non-null   int64  
 13  JobInvolvement            1474 non-null   int64  
 14  JobLevel

In [8]:
df_abc.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1469    False
1470     True
1471     True
1472     True
1473     True
Length: 1474, dtype: bool

## Limpieza

Se detectaron 4 filas duplicadas completas al final del conjunto de datos. Dado que cada registro representa un empleado único y no existe una dimensión temporal en el dataset, estos duplicados no aportan información adicional y se consideran un artefacto del proceso de carga de datos. Por ello, se procedió a su eliminación.

In [9]:
# eliminar duplicados
df_abc_copy = df_abc.copy()
df_abc_copy.drop_duplicates()
df_abc_copy.head().T

0                       1  \
Age                                    41.0                    49.0   
Attrition                               Yes                      No   
BusinessTravel                Travel_Rarely       Travel_Frequently   
DailyRate                              1102                     279   
Department                            Sales  Research & Development   
DistanceFromHome                          1                       8   
Education                                 2                       1   
EducationField                Life Sciences           Life Sciences   
EmployeeCount                             1                       1   
EmployeeNumber                            1                       2   
EnvironmentSatisfaction                   2                       3   
Gender                               Female                    Male   
HourlyRate                               94                      61   
JobInvolvement                            3                       2   
JobLevel                                  2                       2   
JobRole                    sALES eXECUTIVE      rESEARCH sCIENTIST    
JobSatisfaction                         4.0                     2.0   
MaritalStatus                        Single                 Married   
MonthlyIncome                        5993.0                  5130.0   
MonthlyRate                           19479                   24907   
NumCompaniesWorked                        8                       1   
Over18                                    Y                       Y   
OverTime                                Yes                      No   
PercentSalaryHike                        11                      23   
PerformanceRating                         3                       4   
RelationshipSatisfaction                  1                       4   
StandardHours                          80.0                     NaN   
StockOptionLevel                          0                       1   
TotalWorkingYears                         8                      10   
TrainingTimesLastYear                   0.0                     3.0   
WorkLifeBalance                           1                       3   
YearsAtCompany                            6                      10   
YearsInCurrentRole                        4                       7   
YearsSinceLastPromotion                   0                       1   
YearsWithCurrManager                    5.0                     7.0   

                                                2                       3  \
Age                                          37.0                    33.0   
Attrition                                     Yes                      No   
BusinessTravel                      Travel_Rarely       Travel_Frequently   
DailyRate                                    1373                    1392   
Department                 Research & Development  Research & Development   
DistanceFromHome                                2                       3   
Education                                       2                       4   
EducationField                              Other           Life Sciences   
EmployeeCount                                   1                       1   
EmployeeNumber                                  4                       5   
EnvironmentSatisfaction                         4                       4   
Gender                                       Male                  Female   
HourlyRate                                     92                      56   
JobInvolvement                                  2                       3   
JobLevel                                        1                       1   
JobRole                    lABORATORY tECHNICIAN      rESEARCH sCIENTIST    
JobSatisfaction                               3.0                     3.0   
MaritalStatus                              Single                 Married   
MonthlyIncome         

In [10]:
df_abc_copy.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [11]:
df_abc_copy.columns = [re.sub(r'(?<!^)([A-Z])', r'_\1', col).lower() for col in df_abc_copy.columns]
df_abc_copy.columns

Index(['age', 'attrition', 'business_travel', 'daily_rate', 'department',
       'distance_from_home', 'education', 'education_field', 'employee_count',
       'employee_number', 'environment_satisfaction', 'gender', 'hourly_rate',
       'job_involvement', 'job_level', 'job_role', 'job_satisfaction',
       'marital_status', 'monthly_income', 'monthly_rate',
       'num_companies_worked', 'over18', 'over_time', 'percent_salary_hike',
       'performance_rating', 'relationship_satisfaction', 'standard_hours',
       'stock_option_level', 'total_working_years', 'training_times_last_year',
       'work_life_balance', 'years_at_company', 'years_in_current_role',
       'years_since_last_promotion', 'years_with_curr_manager'],
      dtype='object')

In [12]:
# eliminación de columnas innecesarias
# no eliminar employee_count para facilitar el recuento de empleados.
df_abc_copy = df_abc_copy.drop(columns=['over18', 'standard_hours'])

In [13]:
df_abc_copy.columns

Index(['age', 'attrition', 'business_travel', 'daily_rate', 'department',
       'distance_from_home', 'education', 'education_field', 'employee_count',
       'employee_number', 'environment_satisfaction', 'gender', 'hourly_rate',
       'job_involvement', 'job_level', 'job_role', 'job_satisfaction',
       'marital_status', 'monthly_income', 'monthly_rate',
       'num_companies_worked', 'over_time', 'percent_salary_hike',
       'performance_rating', 'relationship_satisfaction', 'stock_option_level',
       'total_working_years', 'training_times_last_year', 'work_life_balance',
       'years_at_company', 'years_in_current_role',
       'years_since_last_promotion', 'years_with_curr_manager'],
      dtype='object')

In [14]:
df_abc_copy.columns = (
    df_abc_copy.columns
      .str.strip()                 # quita espacios al inicio/fin
      .str.lower()
      .str.replace(r"\s+", "_", regex=True)  # cualquier espacio → "_"
)

In [15]:
df_abc_copy['attrition'].unique()

array(['Yes', 'No'], dtype=object)

En visualizacion

categorias de 'work_life_balance' a 1=bad, 2=good, 3=better, 4=best

environment_satisfaction 1=‘Low’ 2=‘Medium’ 3=‘High’ 4=‘Very High’

job_involvement 1=‘Low’ 2=‘Medium’ 3=‘High’ 4=‘Very High’

job_satisfaction 1=‘Low’ 2=‘Medium’ 3=‘High’ 4=‘Very High’

performance_rating 1=‘Low’ 2=‘Good’ 3=‘Excellent’ 4=‘Outstanding’

relationship_satisfaction 1=‘Low’ 2=‘Medium’ 3=‘High’ 4=‘Very High’

In [16]:
columnas_texto = df_abc_copy.select_dtypes(include=['object']).columns

# 2. Aplicamos la "limpieza total" en cada una
for col in columnas_texto:
    # Primero reemplazamos guiones y guiones bajos por espacios
    # Luego quitamos espacios sobrantes en los bordes con .strip()
    # Y finalmente ponemos la primera letra de cada palabra en mayúscula
    df_abc_copy[col] = df_abc_copy[col].str.replace('-', ' ').str.replace('_', ' ').str.strip().str.title()

print("¡Limpieza de contenido completada!")

¡Limpieza de contenido completada!


In [17]:
df_abc_copy['job_role'].unique()

array(['Sales Executive', 'Research Scientist', 'Laboratory Technician',
       'Manufacturing Director', 'Healthcare Representative', 'Manager',
       'Sales Representative', 'Research Director', 'Human Resources'],
      dtype=object)

In [18]:
df_abc_copy['business_travel'].unique()

array(['Travel Rarely', 'Travel Frequently', 'Non Travel', nan],
      dtype=object)

In [20]:
# Creamos el diccionario con el mapeo de los niveles
niveles_educacion = {
    1: 'Below College',
    2: 'College',
    3: 'Bachelor',
    4: 'Master',
    5: 'Doctor'
}

# Aplicamos 
df_abc_copy['education'] = df_abc_copy['education'].replace(niveles_educacion)

# Verificamos el cambio
print(df_abc_copy['education'].unique())

['College' 'Below College' 'Master' 'Bachelor' 'Doctor']


In [21]:
df_abc_copy["age"] = df_abc_copy["age"].astype("Int64")

In [22]:
df_abc_copy["years_with_curr_manager"] = df_abc_copy["years_with_curr_manager"].astype("Int64")

In [23]:
df_abc_copy["training_times_last_year"] = df_abc_copy["training_times_last_year"].astype("Int64")

In [24]:
df_abc_copy["job_satisfaction"] = df_abc_copy["job_satisfaction"].astype("Int64")

In [ ]:
df_abc_copy.info()

In [26]:
df_abc_copy['marital_status'] = df_abc_copy['marital_status'].str.replace('Marreid', 'Married')

In [27]:
df_abc_copy['marital_status'].unique()

array(['Single', 'Married', 'Divorced', nan], dtype=object)

## Gestión de nulos

In [29]:
df_abc_copy.isnull().sum()

age                            73
attrition                       0
business_travel               117
daily_rate                      0
department                     29
distance_from_home              0
education                       0
education_field                58
employee_count                  0
employee_number                 0
environment_satisfaction        0
gender                          0
hourly_rate                     0
job_involvement                 0
job_level                       0
job_role                        0
job_satisfaction               29
marital_status                132
monthly_income                 14
monthly_rate                    0
num_companies_worked            0
over_time                      44
percent_salary_hike             0
performance_rating              0
relationship_satisfaction       0
stock_option_level              0
total_working_years             0
training_times_last_year       88
work_life_balance               0
years_at_compa

In [30]:
# % nulos
nulos = df_abc_copy.isnull().sum()/df_abc_copy.shape[0]*100
# % nulos ordenados de mayor a menor
nulos.sort_values(ascending=False)

years_with_curr_manager       10.040706
marital_status                 8.955224
business_travel                7.937585
training_times_last_year       5.970149
age                            4.952510
education_field                3.934871
over_time                      2.985075
department                     1.967436
job_satisfaction               1.967436
monthly_income                 0.949796
employee_count                 0.000000
gender                         0.000000
employee_number                0.000000
daily_rate                     0.000000
distance_from_home             0.000000
education                      0.000000
attrition                      0.000000
job_role                       0.000000
job_level                      0.000000
job_involvement                0.000000
hourly_rate                    0.000000
environment_satisfaction       0.000000
percent_salary_hike            0.000000
num_companies_worked           0.000000
monthly_rate                   0.000000


years_with_curr_manager       10.040706 - media o mediana
marital_status                 8.955224 - moda
business_travel                7.937585 - asumimos que no viajan ok
training_times_last_year       5.970149 - asuminos que no han tenido formacion ok
age                            4.952510 - media o mediana
education_field                3.934871 - asumimos que va en "otros" ok
over_time                      2.985075 - asumimos que es un "no" ok
department                     1.967436 - moda
job_satisfaction               1.967436 - media o mediana
monthly_income                 0.949796 - media o mediana


In [ ]:
# Gestion nulos business_travel  
df_abc_copy['business_travel'].fillna('Non Travel', inplace=True)

C:\Users\rocio\AppData\Local\Temp\ipykernel_19256\836252022.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_abc_copy['business_travel'].fillna('Non Travel', inplace=True)


In [32]:
df_abc_copy['business_travel'].unique()

array(['Travel Rarely', 'Travel Frequently', 'Non Travel'], dtype=object)

In [33]:
# Gestion nulos training_times_last_year
df_abc_copy['training_times_last_year'].fillna(0, inplace=True)
df_abc_copy['training_times_last_year'].unique()

C:\Users\rocio\AppData\Local\Temp\ipykernel_19256\408558331.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_abc_copy['training_times_last_year'].fillna(0, inplace=True)


<IntegerArray>
[0, 3, 2, 5, 1, 4, 6]
Length: 7, dtype: Int64

In [ ]:
# Gestion nulos over_time: asumimos que es "no" porque si la haces te asugunas de dejar constancia para que las valoren económicamente.
df_abc_copy['over_time'].fillna('No', inplace=True)

In [35]:
df_abc_copy['over_time'].unique()

array(['Yes', 'No'], dtype=object)

In [36]:
# Gestión nulos de education_field
df_abc_copy['education_field'].fillna('Other', inplace=True)

C:\Users\rocio\AppData\Local\Temp\ipykernel_19256\654809224.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_abc_copy['education_field'].fillna('Other', inplace=True)


In [37]:
df_abc_copy['education_field'].unique()

array(['Life Sciences', 'Other', 'Medical', 'Marketing',
       'Technical Degree', 'Human Resources'], dtype=object)

In [ ]:
# Gestión de nulos marital_status: imputamos la moda
#moda_marital_status = df_abc_copy['marital_status'].mode()[0]
#df_abc_copy['marital_status'].fillna(moda_marital_status, inplace=True)